### PR TITLE
feat(dyn-sampling): Apply actual sample_rate to not oversample/undersample

### PR DESCRIPTION
**Pull request by andriisoldatenko**
This PR implements logic to reduce the deviation between the actual sample rate for the organization and the blended sample rate (desired sample rate) by sending new factor rule such as:

```
{
    "condition": {"inner": [], "op": "and"},
    "id": 1004,
    "samplingValue": {"type": "factor", "value": <<<factor>>>},
    "type": "trace",
}
```

```
# previous_factor = 1.0 by default
factor = previous_factor * desired_sample_rate / actual_sample_rate
```

### CODE REVIEW

File: src/sentry/quotas/base.py [hunk 1]
```diff
@@ -469,3 +469,4 @@ def get_blended_sample_rate(self, project: "Project") -> Optional[float]:
 
         :param project: The project model.
         """
+        return 0.25
```
Comment (andriisoldatenko) [inline src/sentry/quotas/base.py:472]:
```suggestion
```

File: src/sentry/dynamic_sampling/rules/helpers/prioritise_project.py [hunk 2]
```diff
@@ -18,7 +45,21 @@ def get_prioritise_by_project_sample_rate(project: "Project", default_sample_rat
...
-        cached_sample_rate = None
-    return cached_sample_rate if cached_sample_rate else default_sample_rate
+        adjusted_sample_rate = None
+
+    cache_key_actual_rate = _generate_cache_key_actual_rate(project.organization.id)
+    try:
+        actual_sample_rate = float(redis_client.hget(cache_key_actual_rate, project.id))
+    except (TypeError, ValueError):
+        actual_sample_rate = None
+    if features.has("organizations:ds-apply-actual-sample-rate-to-biases", project.organization):
```
Comment (iambriccardo) [inline src/sentry/dynamic_sampling/rules/helpers/prioritise_project.py:57]: nit: would leave a space here for more readability.

File: src/sentry/dynamic_sampling/models/utils.py [hunk 3]
```diff
@@ -12,6 +12,16 @@ class DSElement:
     new_sample_rate: float = 0.0
 
 
+def actual_sample_rate(count_keep: int, count_drop: int) -> float:
```
[the remark this answers is not in the document]
Comment (andriisoldatenko) [reply src/sentry/dynamic_sampling/models/utils.py:15]: so another name would be as you proposed: `effective_rate` but we called it with @jan-auer `actual` in slack.

**Review by RaduW — COMMENTED** [review]
I think we are ok, except the calculation of the factor, I think it should be 
` desired/actual` .

Also didn't notice a test where we actually test this calculation (maybe calculation of this factor should not be in the task to make it easier to test). I generally try to put all my calculations in utility functions that are easy and fast to test.

Let's double check the calculation tomorrow, otherwise I think it is good to go

File: src/sentry/dynamic_sampling/rules/biases/factor_bias.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        prev_factor = 1.0
+        if features.has(
+            "organizations:ds-apply-actual-sample-rate-to-biases", project.organization
+        ):
+            redis_client = get_redis_client_for_ds()
+            adj_factor_cache_key = generate_cache_key_adj_factor(project.organization.id)
+            try:
+                prev_factor = float(redis_client.hget(adj_factor_cache_key, project.id))
+            except (TypeError, ValueError):
+                pass
```
Comment (RaduW) [inline src/sentry/dynamic_sampling/rules/biases/factor_bias.py:30]: I this case we don't adjust anything i.e.  `prev_factor == 1` 

I would return an empty array ( no point creating a rule that doesn't do anything).


```suggestion
            except (TypeError, ValueError):
                return []
```

File: src/sentry/dynamic_sampling/rules/utils.py [hunk 5]
```diff
@@ -208,3 +210,25 @@ def apply_dynamic_factor(base_sample_rate: float, x: float) -> float:
 def get_redis_client_for_ds() -> Any:
     cluster_key = getattr(settings, "SENTRY_DYNAMIC_SAMPLING_RULES_REDIS_CLUSTER", "default")
     return redis.redis_clusters.get(cluster_key)
+
+
+def generate_cache_key_adj_factor(org_id: int) -> str:
+    return f"ds::o:{org_id}:prioritise_projects:adj_factor"
```
Comment (RaduW) [inline src/sentry/dynamic_sampling/rules/utils.py:216]: NIT: this is the adjustment factor for everything ( not specific to prioritise_projects) so I wouldn't call it
`ds::o:123:prioritise_projects:adj_factor` maybe just  `ds::o:123:rate_rebalance_factor` or something similar.

```suggestion
    return f"ds::o:{org_id}:rate_rebalance_factor"
```

File: src/sentry/dynamic_sampling/rules/utils.py [hunk 6]
```diff
@@ -208,3 +210,25 @@ def apply_dynamic_factor(base_sample_rate: float, x: float) -> float:
...
+    """
+    try:
+        return count_keep / (count_drop + count_keep)
+    except ZeroDivisionError:
+        return 0.0
+
+
+def adjusted_factor(prev_factor: float, actual_rate: float, desired_sample_rate: float):
+    """
+    Calculate adjusted factor based on blended_sample rate (desired_sample_rate) and actual sample rate
```
Comment (RaduW) [inline src/sentry/dynamic_sampling/rules/utils.py:231]: NIT

```suggestion
    Calculates an adjustment factor in order to bring the actual sample rate to the blended_sample rate (i.e. desired_sample_rate)
```

**Review by RaduW — APPROVED** [review]
Looks good.
There are a few minor suggestions that can be applied if you agree with them. 

Good to go!

**Review by iambriccardo — APPROVED** [review]

**Pull request merged**